src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue`2.cs

```diff
@@ -175,12 +220,24 @@ void AddItemsToBatch(IEnumerable<TItem> items)
                 return _updateTask;
         }
 
-        private async ValueTask<TResult> ProcessNextBatchAsync(CancellationToken cancellationToken)
+        private async ValueTask<TResult?> ProcessNextBatchAsync()
         {
-            cancellationToken.ThrowIfCancellationRequested();
+            _entireQueueCancellationToken.ThrowIfCancellationRequested();
             try
             {
-                return await _processBatchAsync(GetNextBatchAndResetQueue(), _cancellationToken).ConfigureAwait(false);
+                var (nextBatch, batchCancellationToken) = GetNextBatchAndResetQueue();
+
+                // We may have no items if the entire batch was canceled (and no new work was added).
+                if (nextBatch.IsEmpty)
+                    return default;
+
+                return await _processBatchAsync(nextBatch, batchCancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException) when (!_entireQueueCancellationToken.IsCancellationRequested)
+            {
+                // Don't bubble up cancellation to the queue for the nested batch cancellation.  Just because we decided
+                // to cancel this batch isn't something that should stop processing further batches.
+                return default;
             }
```

Comment:
@dibarbet open question if i should have this catch block or not.  THe reason to have it is clarity.  Making it clear that batch-cancellation is very localized and doesn't bleed outwards.  

The reason not to have it is that it actually doesn't serve any purpose.  The caller is written with:

```
                // Await the previous item in the task chain in a non-throwing fashion.  Regardless of whether that last
                // task completed successfully or not, we want to move onto the next batch.
                await lastTask.NoThrowAwaitableInternal(captureContext: false);
```

So even if this batch stops with cancellation (Either for the whole queue, or the single batch) the outer code will continue to run without problems.

Thoughts?

Reply:
I don't mind having the catch here.  It makes it clear that cancellation is an expected scenario rather than just something that occurs on shutdown